In [1]:
import pandas as pd

In [6]:
pd.set_option('display.max_columns', None)
df = pd.read_csv('hotel_bookings_data.csv')
df.head()

hotel  is_canceled  lead_time  arrival_date_year arrival_date_month  \
0  Resort Hotel            0        342               2017          September   
1  Resort Hotel            0        737               2017          September   
2  Resort Hotel            0          7               2017          September   
3  Resort Hotel            0         13               2017          September   
4  Resort Hotel            0         14               2017          September   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   
2                        27                          1   
3                        27                          1   
4                        27                          1   

   stays_in_weekend_nights  stays_in_weekdays_nights  adults  children  \
0                        0                         0       2       0.0   
1                        0                         0       2       0.0   
2                        0                         1       1       0.0   
3                        0                         1       1       0.0   
4                        0                         2       2       0.0   

   babies       meal              city market_segment distribution_channel  \
0       0  Breakfast     Kota Denpasar         Direct               Direct   
1       0  Breakfast     Kota Denpasar         Direct               Direct   
2       0  Breakfast  Kabupaten Bangka         Direct               Direct   
3       0  Breakfast  Kabupaten Bangka      Corporate            Corporate   
4       0  Breakfast  Kabupaten Bangka      Online TA                TA/TO   

   is_repeated_guest  previous_cancellations  previous_bookings_not_canceled  \
0                  0                       0                               0   
1                  0                       0                               0   
2                  0                       0                               0   
3                  0                       0                               0   
4                  0                       0                               0   

   booking_changes deposit_type  agent  company  days_in_waiting_list  \
0                3   No Deposit    NaN      NaN                     0   
1                4   No Deposit    NaN      NaN                     0   
2                0   No Deposit    NaN      NaN                     0   
3                0   No Deposit  304.0      NaN                     0   
4                0   No Deposit  240.0      NaN                     0   

  customer_type   adr  required_car_parking_spaces  total_of_special_requests  \
0      Personal   0.0                            0                          0   
1      Personal   0.0                            0                          0   
2      Personal  75.0                            0                          0   
3      Personal  75.0                            0                          0   
4      Personal  98.0                            0                          1   

  reservation_status  
0          Check-Out  
1          Check-Out  
2          Check-Out  
3          Check-Out  
4          Check-Out

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 29 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_weekdays_nights        119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

In [7]:
#cek data kosong
counts = df.isna().sum().sort_values(ascending=False)
#cek persen data kosong
percentages = round(df.isna().mean() * 100, 3).sort_values(ascending=False)
#gabungkan
null_values = pd.concat([counts, percentages], axis=1, keys=["count", "%"])
null_values

count       %
company                         112593  94.307
agent                            16340  13.686
city                               488   0.409
children                             4   0.003
hotel                                0   0.000
is_repeated_guest                    0   0.000
total_of_special_requests            0   0.000
required_car_parking_spaces          0   0.000
adr                                  0   0.000
customer_type                        0   0.000
days_in_waiting_list                 0   0.000
deposit_type                         0   0.000
booking_changes                      0   0.000
previous_bookings_not_canceled       0   0.000
previous_cancellations               0   0.000
market_segment                       0   0.000
distribution_channel                 0   0.000
is_canceled                          0   0.000
meal                                 0   0.000
babies                               0   0.000
adults                               0   0.000
stays_in_weekdays_nights             0   0.000
stays_in_weekend_nights              0   0.000
arrival_date_day_of_month            0   0.000
arrival_date_week_number             0   0.000
arrival_date_month                   0   0.000
arrival_date_year                    0   0.000
lead_time                            0   0.000
reservation_status                   0   0.000

In [11]:
df['meal'].value_counts()

Breakfast     92310
Dinner        14463
No Meal       10650
Undefined      1169
Full Board      798
Name: meal, dtype: int64

# Handling missing value dan nilai kolom yang tidak sesuai

# Kolom kosong:<br>
- company 112593 baris <br>
- agent 16340 baris <br>
- city 488 baris <br>
- children 4 baris <br>

In [15]:
df_clean = df.copy()

In [16]:
df_clean['company'] = df_clean['company'].fillna(0)
df_clean['agent'] = df_clean['agent'].fillna(0)
df_clean['children'] = df_clean['children'].fillna(0)
df_clean['city'] = df_clean['city'].fillna('unknown')

In [17]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 29 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_weekdays_nights        119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119390 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

# Kolom meal :
- 'undefined' 1169 baris

In [18]:
df_clean = df_clean.replace({'meal': {'Undefined':'No Meal'}})

In [19]:
df_clean['meal'].value_counts()

Breakfast     92310
Dinner        14463
No Meal       11819
Full Board      798
Name: meal, dtype: int64

# Dtypes yg tidak sesuai:
- Kolom company seharusnya integer <br>
- Kolom agent seharusnya integer <br>
- Kolom children seharusnya integer

In [20]:
df_clean['company'] = df_clean['company'].astype('int64')
df_clean['agent'] = df_clean['agent'].astype('int64')
df_clean['children'] = df_clean['children'].astype('int64')

In [21]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 29 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_weekdays_nights        119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119390 non-null  int64  
 11  babies                          119390 non-null  int64  
 12  meal            

# Membuang data yg tidak diperlukan <br>
- jumlah tamu untuk setiap pemesanan

In [25]:
df_clean ['total_guest'] = df_clean['adults'] + df_clean['children'] + df_clean['babies']

In [27]:
df_clean[(df_clean['total_guest'] == 0 )]

hotel  is_canceled  lead_time  arrival_date_year  \
2224    Resort Hotel            0          1               2017   
2409    Resort Hotel            0          0               2017   
3181    Resort Hotel            0         36               2017   
3684    Resort Hotel            0        165               2017   
3708    Resort Hotel            0        165               2017   
...              ...          ...        ...                ...   
115029    City Hotel            0        107               2019   
115091    City Hotel            0          1               2019   
116251    City Hotel            0         44               2019   
116534    City Hotel            0          2               2019   
117087    City Hotel            0        170               2019   

       arrival_date_month  arrival_date_week_number  \
2224             December                        41   
2409             December                        42   
3181              January                        47   
3684             February                        53   
3708             February                        53   
...                   ...                       ...   
115029             August                        26   
115091             August                        26   
116251          September                        28   
116534          September                        28   
117087          September                        30   

        arrival_date_day_of_month  stays_in_weekend_nights  \
2224                            6                        0   
2409                           12                        0   
3181                           20                        1   
3684                           30                        1   
3708                           30                        2   
...                           ...                      ...   
115029                         27                        0   
115091                         30                        0   
116251                         15                        1   
116534                         15                        2   
117087                         27                        0   

        stays_in_weekdays_nights  adults  children  babies       meal  \
2224                           3       0         0       0    No Meal   
2409                           0       0         0       0    No Meal   
3181                           2       0         0       0    No Meal   
3684                           4       0         0       0    No Meal   
3708                           4       0         0       0    No Meal   
...                          ...     ...       ...     ...        ...   
115029                         3       0         0       0  Breakfast   
115091                         1       0         0       0    No Meal   
116251                         1       0         0       0    No Meal   
116534                         5       0         0       0    No Meal   
117087                         2       0         0       0  Breakfast   

                        city market_segment distribution_channel  \
2224           Kota Denpasar      Corporate            Corporate   
2409           Kota Denpasar      Corporate            Corporate   
3181               Kota Batu         Groups                TA/TO   
3684           Kota Denpasar         Groups                TA/TO   
3708           Kota Denpasar         Groups                TA/TO   
...                      ...            ...                  ...   
115029    Kabupaten Sumedang      Online TA                TA/TO   
115091         Kota Denpasar  Complementary               Direct   
116251         Kota Surabaya      Online TA                TA/TO   
116534  Kota Jakarta Selatan      Online TA                TA/TO   
117087    Kota Jakarta Barat  Offline TA/TO                TA/TO   

        is_repeated_guest  previous_cancellations  \
2224                    0                       0   
2409

In [28]:
df_clean ['stay_duration'] = df_clean['stays_in_weekend_nights'] + df_clean['stays_in_weekdays_nights']

In [29]:
df_clean[(df_clean['stay_duration'] == 0 )]

hotel  is_canceled  lead_time  arrival_date_year  \
0       Resort Hotel            0        342               2017   
1       Resort Hotel            0        737               2017   
167     Resort Hotel            0        111               2017   
168     Resort Hotel            0          0               2017   
196     Resort Hotel            0          8               2017   
...              ...          ...        ...                ...   
115483    City Hotel            0         15               2019   
117701    City Hotel            0          0               2019   
118029    City Hotel            0          0               2019   
118631    City Hotel            0         78               2019   
118963    City Hotel            0          1               2019   

       arrival_date_month  arrival_date_week_number  \
0               September                        27   
1               September                        27   
167             September                        28   
168             September                        28   
196             September                        28   
...                   ...                       ...   
115483          September                        27   
117701            October                        32   
118029            October                        33   
118631            October                        34   
118963            October                        35   

        arrival_date_day_of_month  stays_in_weekend_nights  \
0                               1                        0   
1                               1                        0   
167                             6                        0   
168                             6                        0   
196                             7                        0   
...                           ...                      ...   
115483                          6                        0   
117701                          8                        0   
118029                         14                        0   
118631                         23                        0   
118963                         27                        0   

        stays_in_weekdays_nights  adults  children  babies       meal  \
0                              0       2         0       0  Breakfast   
1                              0       2         0       0  Breakfast   
167                            0       2         0       0  Breakfast   
168                            0       1         0       0  Breakfast   
196                            0       2         0       0  Breakfast   
...                          ...     ...       ...     ...        ...   
115483                         0       1         0       0    No Meal   
117701                         0       2         0       0  Breakfast   
118029                         0       2         0       0  Breakfast   
118631                         0       1         0       0  Breakfast   
118963                         0       2         0       0  Breakfast   

                   city market_segment distribution_channel  \
0         Kota Denpasar         Direct               Direct   
1         Kota Denpasar         Direct               Direct   
167       Kota Denpasar      Online TA                TA/TO   
168       Kota Denpasar         Direct               Direct   
196       Kota Denpasar         Direct               Direct   
...                 ...            ...                  ...   
115483  Kota Yogyakarta         Direct               Direct   
117701    Kota Denpasar  Offline TA/TO                TA/TO   
118029    Kota Denpasar         Direct               Direct   
118631    Kota Denpasar  Offline TA/TO                TA/TO   
118963    Kota Denpasar         Direct               Direct   

        is_repeated_guest  previous_cancellations  \
0                       0                       0   
1                       0                       0   
167        

In [30]:
df_hotel = df_clean [(df_clean['total_guest'] > 0) & (df_clean['stay_duration'] > 0)]

In [31]:
df_hotel.head()

hotel  is_canceled  lead_time  arrival_date_year arrival_date_month  \
2  Resort Hotel            0          7               2017          September   
3  Resort Hotel            0         13               2017          September   
4  Resort Hotel            0         14               2017          September   
5  Resort Hotel            0         14               2017          September   
6  Resort Hotel            0          0               2017          September   

   arrival_date_week_number  arrival_date_day_of_month  \
2                        27                          1   
3                        27                          1   
4                        27                          1   
5                        27                          1   
6                        27                          1   

   stays_in_weekend_nights  stays_in_weekdays_nights  adults  children  \
2                        0                         1       1         0   
3                        0                         1       1         0   
4                        0                         2       2         0   
5                        0                         2       2         0   
6                        0                         2       2         0   

   babies       meal              city market_segment distribution_channel  \
2       0  Breakfast  Kabupaten Bangka         Direct               Direct   
3       0  Breakfast  Kabupaten Bangka      Corporate            Corporate   
4       0  Breakfast  Kabupaten Bangka      Online TA                TA/TO   
5       0  Breakfast  Kabupaten Bangka      Online TA                TA/TO   
6       0  Breakfast     Kota Denpasar         Direct               Direct   

   is_repeated_guest  previous_cancellations  previous_bookings_not_canceled  \
2                  0                       0                               0   
3                  0                       0                               0   
4                  0                       0                               0   
5                  0                       0                               0   
6                  0                       0                               0   

   booking_changes deposit_type  agent  company  days_in_waiting_list  \
2                0   No Deposit      0        0                     0   
3                0   No Deposit    304        0                     0   
4                0   No Deposit    240        0                     0   
5                0   No Deposit    240        0                     0   
6                0   No Deposit      0        0                     0   

  customer_type    adr  required_car_parking_spaces  \
2      Personal   75.0                            0   
3      Personal   75.0                            0   
4      Personal   98.0                            0   
5      Personal   98.0                            0   
6      Personal  107.0                            0   

   total_of_special_requests reservation_status  total_guest  stay_duration  
2                          0          Check-Out            1              1  
3                          0          Check-Out            1              1  
4                          1          Check-Out            2              2  
5                          1          Check-Out            2              2  
6                          0          Check-Out            2              2

In [32]:
df_hotel.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 118565 entries, 2 to 119389
Data columns (total 31 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           118565 non-null  object 
 1   is_canceled                     118565 non-null  int64  
 2   lead_time                       118565 non-null  int64  
 3   arrival_date_year               118565 non-null  int64  
 4   arrival_date_month              118565 non-null  object 
 5   arrival_date_week_number        118565 non-null  int64  
 6   arrival_date_day_of_month       118565 non-null  int64  
 7   stays_in_weekend_nights         118565 non-null  int64  
 8   stays_in_weekdays_nights        118565 non-null  int64  
 9   adults                          118565 non-null  int64  
 10  children                        118565 non-null  int64  
 11  babies                          118565 non-null  int64  
 12  meal            